File: src/main/java/de/danielmaile/aether/listeners/ListenerAetherVoid.java
package de.danielmaile.aether.listeners;

import io.papermc.paper.event.entity.EntityMoveEvent;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockFromToEvent;
import org.bukkit.event.player.PlayerMoveEvent;

import static de.danielmaile.aether.worldgen.AetherWorld.getWorld;

public class ListenerAetherVoid implements Listener
{
    /***
     * Teleports players back to the overworld when they fall out of the aether
     */
    @EventHandler
    public void onPlayerMove(PlayerMoveEvent event)
    {
        Player player = event.getPlayer();
        Location playerLoc = player.getLocation();
        if (!player.getWorld().equals(getWorld())) return;
        if (playerLoc.getY() > 64 || playerLoc.getY() < -256) return;

        Location destination = new Location(Bukkit.getWorlds().get(0), playerLoc.getX(), 400, playerLoc.getZ());
        player.teleport(destination);
        event.setCancelled(true);
    }

    /***
     * Kill all entities that fall out of the aether
     */
    @EventHandler
    public void onEntityMove(EntityMoveEvent event)
    {
        Entity entity = event.getEntity();
        if (!entity.getWorld().equals(getWorld())) return;
        if (entity.getLocation().getY() > 64 || entity.getLocation().getY() < -256) return;

        entity.remove();
        event.setCancelled(true);
    }

    /***
     * Stop water and lava from flowing out of aether
     */
    @EventHandler
    public void onBlockFlow(BlockFromToEvent event)
    {
        if (!event.getBlock().getWorld().equals(getWorld())) return;
        if (event.getToBlock().getY() > 64 || event.getToBlock().getY() < -256) return;
        event.setCancelled(true);
    }
}

File: src/main/java/de/danielmaile/aether/listeners/ListenerAetherVoid.kt
package de.danielmaile.aether.listeners

import de.danielmaile.aether.inst
import io.papermc.paper.event.entity.EntityMoveEvent
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.entity.Entity
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockFromToEvent
import org.bukkit.event.player.PlayerMoveEvent

class ListenerAetherVoid : Listener {

    /***
     * Teleports players back to the overworld when they fall out of the aether
     */
    @EventHandler
    fun onPlayerMove(event: PlayerMoveEvent) {
        val player = event.player
        val playerLoc = player.location
        if (player.world != inst().aetherWorld.getWorld()) return
        if (playerLoc.y > 64 || playerLoc.y < -256) return

        val destination = Location(Bukkit.getWorlds()[0], playerLoc.x, 400.0, playerLoc.z)
        player.teleport(destination)
        event.isCancelled = true
    }

    /***
     * Kill all entities that fall out of the aether
     */
    @EventHandler
    fun onEntityMove(event: EntityMoveEvent) {
        val entity: Entity = event.entity
        if (entity.world != inst().aetherWorld.getWorld()) return
        if (entity.location.y > 64 || entity.location.y < -256) return

        entity.remove()
        event.isCancelled = true
    }

    /***
     * Stop water and lava from flowing out of aether
     */
    @EventHandler
    fun onBlockFlow(event: BlockFromToEvent) {
        if (event.block.world != inst().aetherWorld.getWorld()) return
        if (event.toBlock.y > 64 || event.toBlock.y < -256) return
        event.isCancelled = true
    }
}